lib/compress/zstd_double_fast.c

```diff
@@ -48,10 +48,216 @@ void ZSTD_fillDoubleHashTable(ZSTD_matchState_t* ms,
 
 
 FORCE_INLINE_TEMPLATE
-size_t ZSTD_compressBlock_doubleFast_generic(
+size_t ZSTD_compressBlock_doubleFast_noDict_generic(
+        ZSTD_matchState_t* ms, seqStore_t* seqStore, U32 rep[ZSTD_REP_NUM],
+        void const* src, size_t srcSize, U32 const mls /* template */)
+{
+    ZSTD_compressionParameters const* cParams = &ms->cParams;
+    U32* const hashLong = ms->hashTable;
+    const U32 hBitsL = cParams->hashLog;
+    U32* const hashSmall = ms->chainTable;
+    const U32 hBitsS = cParams->chainLog;
+    const BYTE* const base = ms->window.base;
+    const BYTE* const istart = (const BYTE*)src;
+    const BYTE* anchor = istart;
+    const U32 endIndex = (U32)((size_t)(istart - base) + srcSize);
+    /* presumes that, if there is a dictionary, it must be using Attach mode */
+    const U32 prefixLowestIndex = ZSTD_getLowestPrefixIndex(ms, endIndex, cParams->windowLog);
+    const BYTE* const prefixLowest = base + prefixLowestIndex;
+    const BYTE* const iend = istart + srcSize;
+    const BYTE* const ilimit = iend - HASH_READ_SIZE;
+    U32 offset_1=rep[0], offset_2=rep[1];
+    U32 offsetSaved = 0;
+
+    size_t mLength;
+    U32 offset;
+    U32 curr;
+
+    const size_t kStepIncr = 1 << kSearchStrength;
+    const BYTE* nextStep;
+    size_t step;
+
+    size_t hl0;
+    size_t hs0;
+    size_t hl1;
+
+    U32 idxl0;
+    U32 idxs0;
+    U32 idxl1;
+
+    const BYTE* matchl0;
+    const BYTE* matchs0;
+    const BYTE* matchl1;
+
+    const BYTE* ip = istart;
+    const BYTE* ip1;
+
+    DEBUGLOG(5, "ZSTD_compressBlock_doubleFast_noDict_generic");
+
+    /* init */
+    ip += ((ip - prefixLowest) == 0);
+    {
+        U32 const current = (U32)(ip - base);
+        U32 const windowLow = ZSTD_getLowestPrefixIndex(ms, current, cParams->windowLog);
+        U32 const maxRep = current - windowLow;
+        if (offset_2 > maxRep) offsetSaved = offset_2, offset_2 = 0;
+        if (offset_1 > maxRep) offsetSaved = offset_1, offset_1 = 0;
+    }
+
+_start:
```

Comment:
Could it be written as a loop, rather than a `goto` ?

Reply:
Yeah, I thought about it. To a rough approximation, the original implementation looks like this:

```c
size_t dfast() {
  while (ip < ilimit) {
    if (search()) goto _match;
    continue;
_match:
    store();
  }
  return;
}
```

This PR changes it to something like this:

```c
size_t dfast() {
_start:
  if (ip >= ilimit) goto _cleanup;
  init();
  do {
    if (search()) goto _match;
  } while (ip < ilimit);
_cleanup:
  return;
_match:
  store();
  goto _start;
}
```

Admittedly, this abuses `goto`s a bunch. It is quite close to the assembly that gets generated though, which helps me think about the flow.

I believe it could be rewritten to instead look like this:

```c
size_t dfast() {
  if (ip < ilimit) {
    init();
    do {
      if (search()) goto _match;
      continue;
_match:
      store();
      init();
    } while (ip < ilimit);
  }
  return;
}
```

This solves the abuse of `goto`, but has two `init()` blocks, which I don't like. I'm also somewhat attracted to moving the match code outside of the tight loop because it's then clearer what the hot loop is.

Alternatively, it could maybe be structured like this, which only has one `init()` block, but has two loops nested.

I think this would improve the variable scoping concerns.

```c
size_t dfast() {
  while (ip < ilimit) {
    init();
    do {
      if (search()) goto _match;
    } while (ip < ilimit);
    break;
_match:
    store();
  }
  return;
}
```

Reply:
I have a version of this last approach [here](https://github.com/felixhandte/zstd/compare/zstd-dfast-pipelined-single...felixhandte:zstd-dfast-pipelined-loops?expand=1). Performance looks good on gcc-10, but it's slower on clang-12. :/